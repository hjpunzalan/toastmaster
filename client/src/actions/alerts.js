import { SET_ALERT, RESET_ALERT } from './types';

export const setAlert = (msg, alertType) => dispatch => {
	dispatch({
		type: SET_ALERT,
		payload: { msg, alertType }
	});
};

export const resetAlert = () => dispatch => {
	dispatch({ type: RESET_ALERT });
};
