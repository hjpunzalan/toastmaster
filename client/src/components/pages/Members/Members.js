import React from 'react';
import { MdEmail } from 'react-icons/md';
import img from '../../../img/anonymous.png';

const List = ({ user }) => {
	return (
		<tr key={user._id} className="MemberList__member">
			<td>
				<img
					className="MemberList__member-photo"
					src={user.photo || img}
					alt="user avatar"
				/>
			</td>
			<td>
				<span className="MemberList__member-name">{user.firstName}</span>
			</td>
			<td>
				<span className="MemberList__member-name">{user.lastName}</span>
			</td>
			<td>
				<span
					style={{
						display: 'flex',
						alignItems: 'center',
						justifyContent: 'center'
					}}>
					<a
						href={`mailto:${user.email}`}
						target="_top"
						className="MemberList__member-email">
						<MdEmail />
					</a>
					{user.email}
				</span>
			</td>
		</tr>
	);
};

export default List;
