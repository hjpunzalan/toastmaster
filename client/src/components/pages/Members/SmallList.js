import React, { useEffect } from 'react';
import { MdEmail } from 'react-icons/md';
import img from '../../../img/anonymous.png';

const SmallList = ({
	users,
	Moderator,
	active,
	deActivateUser,
	activateUser,
	currentUser,
	setActive,
	setAlert,
	resetAlert,
	changeRole
}) => {
	useEffect(() => {
		if (users.length === 0) {
			resetAlert();
			setActive(true);
			setAlert('Currently there are no unactive users', 'fail');
		}
		// eslint-disable-next-line
	}, [users]);

	return (
		<div className="SmallList">
			{users.map(user => (
				<div className="SmallList__user" key={user._id}>
					<img
						className="SmallList__photo"
						src={user.photo || img}
						alt="avatar"
					/>
					<div className="SmallList__body">
						<div className="SmallList__details">
							<span className="SmallList__name">
								{user.firstName}&nbsp;{user.lastName}
							</span>
							{(active || Moderator) && (
								<a
									className="SmallList__email"
									href={`mailto:${user.email}`}
									target="_top">
									<MdEmail />
									<span>{user.email}</span>
								</a>
							)}
						</div>
						{Moderator && (
							<div className="SmallList__moderator">
								{Moderator && currentUser.role === 'admin' && active && (
									<div className="SmallList__committee">
										<strong>Committee</strong>
										<select
											defaultValue={
												user.role === 'committee' || user.role === 'admin'
											}
											onChange={e => changeRole(user._id, e.target.value)}>
											<option value={true}>Yes</option>
											<option value={false}>No</option>
										</select>
									</div>
								)}
								{active ? (
									<div
										className="SmallList__deactivate"
										onClick={() => deActivateUser(user._id)}>
										Deactivate
									</div>
								) : (
									<div
										className="SmallList__activate"
										onClick={() => activateUser(user._id)}>
										Activate
									</div>
								)}
							</div>
						)}
					</div>
				</div>
			))}
		</div>
	);
};

export default SmallList;
