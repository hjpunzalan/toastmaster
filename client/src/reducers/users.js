import {
	REGISTER_SUCCESS,
	GET_ALL_USERS,
	TOGGLE_MODERATOR,
	USER_ERROR,
	DEACTIVATE_USER,
	ACTIVATE_USER,
	LOADING_USER
} from '../actions/types';

const initialState = {
	users: [],
	loading: true,
	error: {},
	Moderator: false
};

export default (state = initialState, action) => {
	const { type, payload } = action;

	const replaceUser = (id, userArray, newUser) => {
		const index = userArray.findIndex(el => el._id === id);
		userArray[index] = newUser;
		return userArray;
	};

	switch (type) {
		case USER_ERROR:
			return { ...state, loading: false };
		case LOADING_USER:
			return { ...state, loading: true };
		case REGISTER_SUCCESS:
			localStorage.setItem(
				'users',
				JSON.stringify([...state.users, ...payload])
			);
			return {
				//State needs to be first
				...state,
				users: [...state.users, ...payload],
				loading: false
			};
		case GET_ALL_USERS:
			return {
				...state,
				loading: false,
				users: payload
			};
		case TOGGLE_MODERATOR:
			return {
				...state,
				Moderator: !state.Moderator
			};

		case DEACTIVATE_USER:
		case ACTIVATE_USER:
			return {
				...state,
				users: replaceUser(payload._id, state.users, payload),
				loading: false
			};

		default:
			return state;
	}
};
