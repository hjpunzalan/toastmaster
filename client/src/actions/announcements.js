import axios from 'axios';
import {
	TOGGLE_CREATE_ANNOUNCEMENT,
	CREATE_ANNOUNCEMENT,
	GET_ALL_ANNOUNCEMENT,
	DELETE_ANNOUNCEMENT,
	UPDATE_ANNOUNCEMENT,
	LOADING_ANNOUNCEMENT_SUBMIT
} from './types';
import { setAlert, resetAlert } from './alerts';
import catchAsync from '../utils/catchAsync';

export const toggleEdit = () => dispatch => {
	dispatch(resetAlert());
	dispatch({ type: TOGGLE_CREATE_ANNOUNCEMENT });
};

export const createAnnouncement = (title, contentState, plainText) =>
	catchAsync('announcement', async dispatch => {
		dispatch(resetAlert()); //Need to be in every post/put/patch action with alert
		const jsonContentState = JSON.stringify(contentState);
		// This makes it more UX friendly calling a spinner instantly
		dispatch({ type: LOADING_ANNOUNCEMENT_SUBMIT });
		// Need to be in edit also
		const body = { title, contentState: jsonContentState, plainText };
		const res = await axios.post('/api/announcements', body, {
			headers: {
				'Content-type': 'application/json'
			}
		});

		dispatch({
			type: CREATE_ANNOUNCEMENT,
			payload: res.data
		});
		dispatch(setAlert('New announcement posted!', 'success'));
	});

export const getAnnouncements = () =>
	catchAsync('announcement', async dispatch => {
		const limit = 10; //limit to 10 docs
		const res = await axios.get(
			`/api/announcements?sort=-lastEdited,-date&limit=${limit}`
		);
		dispatch({ type: GET_ALL_ANNOUNCEMENT, payload: res.data });
	});

export const updateAnnouncement = (id, newTitle, newContentState, plainText) =>
	catchAsync('announcement', async dispatch => {
		dispatch(resetAlert());
		dispatch({ type: LOADING_ANNOUNCEMENT_SUBMIT });
		const jsonContentState = JSON.stringify(newContentState);
		const res = await axios.patch(
			`/api/announcements/${id}`,
			{
				title: newTitle,
				contentState: jsonContentState,
				plainText
			},
			{
				headers: {
					'Content-type': 'application/json'
				}
			}
		);
		dispatch({ type: UPDATE_ANNOUNCEMENT, payload: res.data });
		dispatch(setAlert('Announcement updated', 'success'));
	});

export const deleteAnnouncement = id =>
	catchAsync('announcement', async dispatch => {
		if (window.confirm('Are you sure you want to delete the announcement?')) {
			dispatch(resetAlert());
			dispatch({ type: LOADING_ANNOUNCEMENT_SUBMIT });
			await axios.delete(`/api/announcements/${id}`);
			dispatch({ type: DELETE_ANNOUNCEMENT, payload: id });
			dispatch(setAlert('Announcement removed', 'success'));
		} else return;
	});
