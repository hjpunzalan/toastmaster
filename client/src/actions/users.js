import { REGISTER_SUCCESS } from '../actions/types';
import { setAlert, resetAlert } from './alerts';

export const registerUser = formData => dispatch => {
	const { firstName, lastName, email } = formData;
	dispatch(resetAlert());
	const users = [
		{
			firstName: 'Jonathan',
			lastName: 'Punzalan',
			email: 'hj.punzalan@hotmail.com'
		}
	];

	const checkUsers = users.filter(user => user.email === email);

	if (checkUsers.length === 0) {
		dispatch({
			type: REGISTER_SUCCESS,
			payload: formData
		});

		dispatch(
			setAlert(`${firstName} ${lastName} succesfully registered.`, 'success')
		);
	} else
		dispatch(setAlert(`Email: ${email} has already been registered`, 'fail'));
};
