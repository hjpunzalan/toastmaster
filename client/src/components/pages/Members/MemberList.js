import React, { useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import {
	getAllUsers,
	deActivateUser,
	activateUser
} from '../../../actions/users';
import Table from './Table';
import Spinner from '../../utils/Spinner';

const MemberList = ({
	getAllUsers,
	users: { users, loading, Moderator },
	deActivateUser,
	activateUser
}) => {
	useEffect(() => {
		getAllUsers(active);
		//eslint-disable-next-line
	}, []);

	const [active, setActive] = useState(true);
	const activeUsers = users.filter(user => user.active === true);
	const unActiveUsers = users.filter(user => user.active === false);

	return loading ? (
		<Spinner />
	) : (
		<div className="MemberList">
			<h1 className="MemberList__title">Member List</h1>
			<div className="MemberList__top">
				<Link className="MemberList__register" to="/register">
					{Moderator && <button>Register a User</button>}
				</Link>

				<div className="MemberList__switch">
					<input
						type="radio"
						name="users"
						id="active"
						onChange={() => setActive(!active)}
						checked={active}
					/>
					<label htmlFor="active">Active</label>
					<input
						type="radio"
						name="users"
						id="unactive"
						onChange={() => setActive(!active)}
						checked={!active}
					/>
					<label htmlFor="unactive">Unactive</label>
				</div>
			</div>
			<Table
				Moderator={Moderator}
				users={active ? activeUsers : unActiveUsers}
				active={active}
				deActivateUser={deActivateUser}
				activateUser={activateUser}
			/>
		</div>
	);
};

MemberList.propTypes = {
	getAllUsers: PropTypes.func.isRequired,
	users: PropTypes.object.isRequired,
	deActivateUser: PropTypes.func.isRequired,
	activateUser: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
	users: state.users
});

export default connect(
	mapStateToProps,
	{ getAllUsers, deActivateUser, activateUser }
)(MemberList);
