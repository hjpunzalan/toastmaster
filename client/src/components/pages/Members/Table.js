import React from 'react';
import Members from './Members';

const Table = ({ users, Moderator, active, deActivateUser, activateUser }) => {
	return (
		<table className="MemberList__list">
			<thead>
				<tr>
					<th>Photo</th>
					<th>First Name</th>
					<th>Last Name</th>
					{(Moderator || active) && <th>Email</th>}
					{Moderator && (active ? <th>Deactivate</th> : <th>Activate</th>)}
				</tr>
			</thead>
			<tbody>
				{users.map(user => (
					<Members
						key={user._id}
						user={user}
						Moderator={Moderator}
						active={active}
						deActivateUser={deActivateUser}
						activateUser={activateUser}
					/>
				))}
			</tbody>
		</table>
	);
};

export default Table;
