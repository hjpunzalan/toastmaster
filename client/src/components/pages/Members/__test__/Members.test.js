
import React from "react";
import { BrowserRouter } from "react-router-dom";
import renderer from 'react-test-renderer';
import { shallow } from "enzyme";
import { createBrowserHistory } from "history";
import { deActivateUser,activateUser,changeRole } from "../../../../actions/users";
import { Members } from '../Members'


const history = createBrowserHistory();
jest.spyOn(history, "push");

const initialProps = {
    user:{
                _id: "test",
                 firstName: "test",
            }
        ,
        loading: false,
        Moderator: false
    , active: true,
    deActivateUser,
    activateUser,
        currentUser: {
            active: true,
            role: "user",
            _id: "test",
            firstName: "test",
            lastName: "test",
           email:"user@test.com"
            
    },
    changeRole
}

const setup = (props = {}) => {
	const setupProps = { ...initialProps, ...props };
	return shallow(<Members {...setupProps} />);
};

test('render Members component', () => {
    const wrapper = setup();
    expect(wrapper.find(".MemberList__member").length).toBe(1)
})

test("Snapshot renders correctly", () => {
	const tree = renderer
		.create(<BrowserRouter><Members {...initialProps} /></BrowserRouter>)
		.toJSON();
	expect(tree).toMatchSnapshot()

})
