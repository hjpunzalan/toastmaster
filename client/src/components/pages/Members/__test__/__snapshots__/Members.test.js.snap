// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot renders correctly 1`] = `
<tr
  className="MemberList__member"
>
  <td>
    <img
      alt="user avatar"
      className="MemberList__member-photo"
      src="anonymous.png"
    />
  </td>
  <td>
    <span
      className="MemberList__member-name"
    >
      test
    </span>
  </td>
  <td>
    <span
      className="MemberList__member-name"
    />
  </td>
  <td>
    <span
      style={
        Object {
          "alignItems": "center",
          "display": "flex",
          "justifyContent": "center",
        }
      }
    >
      <a
        className="MemberList__member-email"
        href="mailto:undefined"
        target="_top"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4l-8 5-8-5V6l8 5 8-5v2z"
          />
        </svg>
      </a>
    </span>
  </td>
</tr>
`;
